src/client/api/rpc.c

```diff
@@ -151,8 +164,16 @@ daos_rpc_proto_query(crt_opcode_t base_opc, uint32_t *ver_array, int count, int
 		D_GOTO(out_detach, rc = -DER_NOMEM);
 
 	/** select a random rank to issue the proto query rpc to */
-	rproto->nr_ranks   = dc_mgmt_net_get_num_srv_ranks();
-	rproto->ep.ep_rank = d_rand() % rproto->nr_ranks;
+	nr_ranks = dc_mgmt_net_get_num_srv_ranks();
```

Comment:
I know that you are just improving the existing logic here, but I wonder if it is really necessary to only use the MS ranks for the proto query. If I'm understanding correctly, clients will always do this proto query on startup, so in the case of a large number of clients all starting at the same time, a small number of ranks will have to service all of the proto queries. Maybe it would be better to pick from the entire set of valid ranks in order to spread the load?

Reply:
is this really what is happening here?
nr_ranks is all ranks not just ms ranks. we definitely do not want to just ping the ms ranks.

Reply:
Oops, you're right. I misunderstood the code. The `dc_mgmt_net_get_num_srv_ranks()` helper is returning the total number of ranks in the AttachInfo message. In any case, I think my point about sending the server protocol version in the AttachInfo payload instead of querying from every client still makes sense. But that's a different discussion that I'll bring up elsewhere.